pkg/netpol/eval/check.go

```diff
@@ -270,82 +270,163 @@ func (pe *PolicyEngine) allAllowedConnectionsBetweenPeers(srcPeer, dstPeer Peer)
+	return &policiesLayerXgressConns{isCaptured: false, layerConns: k8s.NewPolicyConnections()}
+}
+
 // allAllowedXgressConnections: returns the allowed connections from srcPeer to dstPeer on the
 // given direction (ingress/egress)
 // allowed conns are computed by walking through all the available resources of k8s network policy api:
 // admin-network-policies, network-policies and baseline-admin-network-policies;
 // considering the precedence of each policy
+// in case of exposure-analysis it also checks and updates if a src is exposed to all namespaces on egress
+// or dst is exposed to all namespaces on ingress
 func (pe *PolicyEngine) allAllowedXgressConnections(src, dst k8s.Peer, isIngress bool) (allowedConns *common.ConnectionSet, err error) {
 	// first get allowed xgress conn between the src and dst from the ANPs
+	// (in case of exposure-analysis get also cluster wide conns of the selected peer from the ANPs)
 	// note that:
-	// - anpConns may contain allowed, denied or/and passed connections
-	// - anpCaptured is true iff there is at least one rule in the input ANPs that captures both src and dst;
+	// - anpConns.layerConns (or anpExposure.layerConns) may contain allowed, denied or/and passed connections;
+	// connections consider rules and policies precedence
+	// - anpConns.isCaptured is true iff there is at least one rule in the input ANPs that captures both src and dst;
 	// because anp rules are read as is and don't contain any implicit isolation effects for the Pods selected by the AdminNetworkPolicy.
-	anpConns, anpCaptured, err := pe.getAllAllowedXgressConnectionsFromANPs(src, dst, isIngress)
+	// anpExposure - is relevant only if pe.exposureAnalysisFlag == true
+	// - anpExposure.isCaptured is true iff there is at least one ANP that exposes the relevant peer (src on egress/ dst on ingress)
+	// to entire-cluster
+	anpConns, anpExposure, err := pe.getAllAllowedXgressConnectionsFromANPs(src, dst, isIngress)
 	if err != nil {
 		return nil, err
 	}
-	// optimization: if all the conns between src and dst were determined by the ANPs : return the allowed conns
-	if anpCaptured && anpConns.DeterminesAllConns() {
-		return anpConns.AllowedConns, nil
+	// optimizations
+	podExposureUpdatedFlag := false
+	anpConnsDeterminedAllFlag := false
+	// if all cluster-wide conns were determined by the ANP then update the selected peer's cluster wide exposure
+	if pe.exposureAnalysisFlag && anpExposure.layerConns.DeterminesAllConns() {
+		podExposureUpdatedFlag = true
+		updatePeerXgressClusterWideExposure(anpExposure.layerConns.AllowedConns, src, dst, isIngress)
+	}
+	// if all the conns between src and dst were determined by the ANPs : return the allowed conns
+	if anpConns.layerConns.DeterminesAllConns() {
+		if !pe.exposureAnalysisFlag || podExposureUpdatedFlag {
+			return anpConns.layerConns.AllowedConns, nil
+		}
+		// else : exposure-analysis is on and still not determined continue
+		anpConnsDeterminedAllFlag = true
```

Comment:
how can get here?

Reply:
because if a->b allows all (or at least determined all conns (ports) from ANPs) does not necessary mean that all exposure to entire cluster was also determined from ANPs .. so need to analyze also NPs / BANP for exposure

Reply:
not sure I follow why

Reply:
example:
ANP denies all conns from a -> b  (ANP determined all conns to b; if exposure analysis is off no need to continue to NP/BANP)
But:
NP exposes A to entire-cluster on all-conns (if we don't continue - this would not be computed and will have wrong results as exposure to entire-cluster is computed once for each direction)

Reply:
ok. might be useful to add this example in comment as documentation, so it is clear.

Reply:
done